okio/src/jvmMain/kotlin/okio/AsyncTimeout.kt

```diff
@@ -195,17 +149,35 @@ open class AsyncTimeout : Timeout() {
   }
 
   /**
-   * Returns either `cause` or an IOException that's caused by `cause` if a timeout
+   * Returns either [cause] or an IOException that's caused by [cause] if a timeout
    * occurred. See [newTimeoutException] for the type of exception
    * returned.
    */
   internal fun exit(cause: IOException): IOException {
     return if (!exit()) cause else newTimeoutException(cause)
   }
 
+  /**
+   * Surrounds [block] with calls to [enter] and [exit], throwing an exception from
+   * [newTimeoutException] if a timeout occurred.
+   */
+  internal inline fun <T> withTimeout(block: () -> T): T {
```

Comment:
Love this

Reply:
Should we make this public?

Reply:
(Lets do the public/internal decision in follow-up)

Reply:
If this is made public, as an inline method, the 2 `exit` methods it calls will also need to be made public. Or they can be inlined into `withTimeout` as that is the only place they are now used.

Is it common for users to use `enter` and `exit` manually? Or is use of the prebuilt `source` and `sink` methods more common? It looks like OkHttp uses `enter` and `exit` in a couple different places; would OkHttp benefit from a `withTimeout` method?

Reply:
Definitely don’t want too many exit methods on the public API. If everyone can use withTimeout() that’s ideal.

Whatever we do I think we’d want to do it so that OkHttp could take advantage.